crates/emitter/src/emitter.rs

```diff
@@ -317,11 +410,101 @@ impl Emitter {
             self.range(&rhs.range);
             self.str("}");
         } else {
+            self.inside_element_opration(lhs, rhs);
+        }
+    }
+
+    fn inside_normal_expression(&mut self, arg: &InsideExpression) {
+        self.str("((");
+        self.expression(&arg.expression);
+        self.str(")");
+        self.space(1);
+        self.inside(&arg.inside);
+        self.space(1);
+        self.l_brace(&arg.l_brace);
+        self.range_list(&arg.range_list);
+        self.r_brace(&arg.r_brace);
+        self.str(")");
+    }
+
+    fn inside_expanded_expression(&mut self, arg: &InsideExpression) {
+        self.str("(");
+        self.inside_element_opration(&arg.expression, &arg.range_list.range_item);
+        for x in &arg.range_list.range_list_list {
+            self.space(1);
+            self.str("||");
+            self.space(1);
+            self.inside_element_opration(&arg.expression, &x.range_item);
+        }
+        self.str(")");
+    }
+
+    fn outside_normal_expression(&mut self, arg: &OutsideExpression) {
+        self.str("!((");
+        self.expression(&arg.expression);
+        self.str(")");
+        self.space(1);
+        self.token(&arg.outside.outside_token.replace("inside"));
+        self.space(1);
+        self.l_brace(&arg.l_brace);
+        self.range_list(&arg.range_list);
+        self.r_brace(&arg.r_brace);
+        self.str(")");
+    }
+
+    fn outside_expanded_expression(&mut self, arg: &OutsideExpression) {
+        self.str("!(");
+        self.inside_element_opration(&arg.expression, &arg.range_list.range_item);
+        for x in &arg.range_list.range_list_list {
+            self.space(1);
+            self.str("||");
+            self.space(1);
+            self.inside_element_opration(&arg.expression, &x.range_item);
+        }
+        self.str(")");
+    }
+
+    fn inside_element_opration(&mut self, lhs: &Expression, rhs: &RangeItem) {
+        if let Some(ref x) = rhs.range.range_opt {
+            let (op_l, op_r) = match &*x.range_operator {
+                RangeOperator::DotDot(_) => (">=", "<"),
+                RangeOperator::DotDotEqu(_) => (">=", "<="),
+            };
+            self.str("(");
+            self.str("(");
+            self.expression(lhs);
+            self.str(")");
+            self.space(1);
+            self.str(op_l);
+            self.space(1);
+            self.str("(");
+            self.expression(&rhs.range.expression);
+            self.str(")");
+            self.str(")");
+            self.space(1);
+            self.str("&&");
+            self.space(1);
+            self.str("(");
+            self.str("(");
+            self.expression(lhs);
+            self.str(")");
+            self.space(1);
+            self.str(op_r);
+            self.space(1);
+            self.str("(");
+            self.expression(&x.expression);
+            self.str(")");
+            self.str(")");
```

Comment:
This sequence seems to be bit long.
I think merging some `str` and `space` into a `str` is no problem. For example, `self.str(")) && ((")`.

Reply:
I fixed your feedback.